### PR TITLE
GPR_LIKELY annotations for client_channel

### PR DESCRIPTION
**Pull request by yashykt**
Client channel GPR_LIKELY annotations based on @markdroth 's recommendations in #15044

---
This change is [Reviewable](https://reviewable.io/reviews/grpc/grpc/15365)

### CODE REVIEW

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE                                                             FILE SIZE
 ++++++++++++++ GROWING                                               ++++++++++++++
  +0.4%    +112 src/core/ext/filters/client_channel/client_channel.cc    +112  +0.4%
      +7.4%     +73 pick_callback_start_locked                                +73  +7.4%
      +3.4%     +27 pick_after_resolver_result_done_locked                    +27  +3.4%
      +6.3%     +22 add_retriable_send_initial_metadata_op                    +22  +6.3%
      +2.8%     +16 cc_destroy_call_elem                                      +16  +2.8%
      +1.1%      +9 pick_done                                                  +9  +1.1%
      +2.3%      +8 recv_message_ready                                         +8  +2.3%
      +2.3%      +8 recv_initial_metadata_ready                                +8  +2.3%
      +0.1%      +3 on_complete                                                +3  +0.1%
      +0.2%      +1 [Unmapped]                                                 +1  +0.2%

 -------------- SHRINKING                                             --------------
  -0.0%      -8 [None]                                                   -872  -0.0%

  +0.0%    +104 TOTAL                                                    -760  -0.0%


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
[microbenchmarks] Performance differences noted:
Benchmark                                                                              cpu_time    real_time
-------------------------------------------------------------------------------------  ----------  -----------
BM_StreamingPingPongMsgs<MinInProcess, NoOpMutator, NoOpMutator>/262144                +7%         +7%
BM_StreamingPingPongWithCoalescingApi<InProcess, NoOpMutator, NoOpMutator>/262144/2/0  +7%         +7%
```

**Comment by markdroth** [comment]
I'd be curious to know how this impacts our benchmarks.

---

Reviewed 1 of 1 files at r1.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks failed.

---

*[src/core/ext/filters/client_channel/client_channel.cc, line 2992 at r1](https://reviewable.io/reviews/grpc/grpc/15365#-LD3j63-8RVrTVBPrQoM:-LD3j63-8RVrTVBPrQoN:b-g55dpg) ([raw file](https://github.com/grpc/grpc/blob/a8ca226fb7a1759743db209bebe5e2bb4ad28e81/src/core/ext/filters/client_channel/client_channel.cc#L2992)):*
> ```cpp
>   }
>   // Handle cancellation.
>   if (batch->cancel_stream) {
> ```

This can use `GPR_UNLIKELY()`.

---

*[src/core/ext/filters/client_channel/client_channel.cc, line 3071 at r1](https://reviewable.io/reviews/grpc/grpc/15365#-LD3jJBA8IzfNMlql3Lp:-LD3jJBA8IzfNMlql3Lq:bithrlv) ([raw file](https://github.com/grpc/grpc/blob/a8ca226fb7a1759743db209bebe5e2bb4ad28e81/src/core/ext/filters/client_channel/client_channel.cc#L3071)):*
> ```cpp
>   calld->owning_call = args->call_stack;
>   calld->call_combiner = args->call_combiner;
>   if (chand->deadline_checking_enabled) {
> ```

This can use `GPR_LIKELY()`.

---


*Comments from [Reviewable](https://reviewable.io/reviews/grpc/grpc/15365#-:-LD3jSUt3p4ObbbVVViP:b1ve4mh)*

File: src/core/ext/filters/client_channel/client_channel.cc [hunk 1]
```diff
@@ -3094,7 +3100,7 @@ static void cc_destroy_call_elem(grpc_call_element* elem,
   for (size_t i = 0; i < GPR_ARRAY_SIZE(calld->pending_batches); ++i) {
     GPR_ASSERT(calld->pending_batches[i].batch == nullptr);
   }
-  if (calld->pick.connected_subchannel != nullptr) {
+  if (GPR_LIKELY(calld->pick.connected_subchannel != nullptr)) {
     calld->pick.connected_subchannel.reset();
   }
   for (size_t i = 0; i < GRPC_CONTEXT_COUNT; ++i) {
```
Comment (yashykt) [inline src/core/ext/filters/client_channel/client_channel.cc]: Some values can be empty, for example, for non-traced, insecure calls

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE                                                             FILE SIZE
 ++++++++++++++ GROWING                                               ++++++++++++++
  +0.0%     +16 [None]                                                +2.95Ki  +0.0%
  +0.3%     +96 src/core/ext/filters/client_channel/client_channel.cc     +96  +0.3%
      +6.8%     +67 pick_callback_start_locked                                +67  +6.8%
      +3.4%     +27 pick_after_resolver_result_done_locked                    +27  +3.4%
      +6.3%     +22 add_retriable_send_initial_metadata_op                    +22  +6.3%
      +1.1%      +9 pick_done                                                  +9  +1.1%
      +2.3%      +8 recv_message_ready                                         +8  +2.3%
      +2.3%      +8 recv_initial_metadata_ready                                +8  +2.3%
      +1.6%      +7 [Unmapped]                                                 +7  +1.6%
      +0.1%      +2 cc_start_transport_stream_op_batch                         +2  +0.1%

  +0.0%    +112 TOTAL                                                 +3.05Ki  +0.0%


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by yashykt** [comment]
Added the other recommendations too except for one. (Not sure if they should be made though, because this means consciously saying that doing things like a single rpc, cancel_batch and disabling deadlines is meant to be slow.

**Comment by yashykt** [comment]
I've also run benchmark runs on this. I should have numbers in about 3 hours.

**Comment by grpc-testing** [comment]
```
[microbenchmarks] No significant performance differences
```

**Comment by markdroth** [comment]
Reviewed 1 of 1 files at r3.
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[src/core/ext/filters/client_channel/client_channel.cc, line 3106 at r1](https://reviewable.io/reviews/grpc/grpc/15365#-LD3hST_Nb_h3AQsXXq8-r1-3106:-LD7CxgR0ZuaxtVWDmJC:bexc4bz) ([raw file](https://github.com/grpc/grpc/blob/a8ca226fb7a1759743db209bebe5e2bb4ad28e81/src/core/ext/filters/client_channel/client_channel.cc#L3106)):*
<details><summary><i>Previously, yashykt (Yash Tibrewal) wrote…</i></summary><blockquote>

Some values can be empty, for example, for non-traced, insecure calls
</blockquote></details>

I think you're right on this one -- I would expect the security context to be non-null in the common case, but the others are probably not used as often.

---

*[src/core/ext/filters/client_channel/client_channel.cc, line 2955 at r3](https://reviewable.io/reviews/grpc/grpc/15365#-LD7CjGkBhc6PSsSrLjI:-LD7CjGkBhc6PSsSrLjJ:b7zufej) ([raw file](https://github.com/grpc/grpc/blob/cff3f9fd150df8b7db5d169303733bcffda4bd46/src/core/ext/filters/client_channel/client_channel.cc#L2955)):*
> ```cpp
>       return;
>     }
>     if (GPR_LIKELY(!chand->started_resolving)) {
> ```

This should be `GPR_UNLIKELY()`.  It will only be true when the channel is idle.

---


*Comments from [Reviewable](https://reviewable.io/reviews/grpc/grpc/15365)*

**Comment by yashykt** [comment]
These were the results from -
On HEAD 
```
+--------+---------+---------+----------------------+----------------------+
|        |  l_50   |  l_99   |  server_qps_per_cpu  |  client_qps_per_cpu  |
+========+=========+=========+======================+======================+
| AVG    | 63.5064 | 126.705 |       41661.9        |       38646.3        |
+--------+---------+---------+----------------------+----------------------+
| STDEV  | 5.52187 | 8.22963 |         2105         |       3188.86        |
+--------+---------+---------+----------------------+----------------------+
| MEDIAN | 64.1037 | 127.629 |        41400         |       38380.4        |
+--------+---------+---------+----------------------+----------------------+
```

With GPR_LIKELY change
```
+--------+---------+---------+----------------------+----------------------+
|        |  l_50   |  l_99   |  server_qps_per_cpu  |  client_qps_per_cpu  |
+========+=========+=========+======================+======================+
| AVG    | 62.8346 | 125.397 |       41505.9        |       39647.9        |
+--------+---------+---------+----------------------+----------------------+
| STDEV  | 5.43443 | 8.2653  |        3150.9        |       2884.49        |
+--------+---------+---------+----------------------+----------------------+
| MEDIAN | 62.3697 | 127.039 |       41657.1        |       39726.4        |
+--------+---------+---------+----------------------+----------------------+
```

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE                                                             FILE SIZE
 ++++++++++++++ GROWING                                               ++++++++++++++
  [ = ]       0 [None]                                                +3.07Ki  +0.0%
  +0.3%     +80 src/core/ext/filters/client_channel/client_channel.cc     +80  +0.3%
      +6.8%     +67 pick_callback_start_locked                                +67  +6.8%
      +3.4%     +27 pick_after_resolver_result_done_locked                    +27  +3.4%
      +6.3%     +22 add_retriable_send_initial_metadata_op                    +22  +6.3%
      +1.1%      +9 pick_done                                                  +9  +1.1%
      +2.3%      +8 recv_message_ready                                         +8  +2.3%
      +2.3%      +8 recv_initial_metadata_ready                                +8  +2.3%
      +0.1%      +2 cc_start_transport_stream_op_batch                         +2  +0.1%

  +0.0%     +80 TOTAL                                                 +3.15Ki  +0.0%


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
[microbenchmarks] Performance differences noted:
Benchmark                                                               cpu_time    real_time
----------------------------------------------------------------------  ----------  -----------
BM_StreamingPingPongMsgs<MinInProcess, NoOpMutator, NoOpMutator>/32768  -4%         -4%
```

**Comment by markdroth** [comment]
Looks like this does yield a noticable improvement.  Thanks for doing this!

---

Reviewed 1 of 1 files at r4.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io/reviews/grpc/grpc/15365#-:-LDHeik53Cq6nCXMy7vF:b-ddbhqs)*

**Comment by yashykt** [comment]
Known issues : #15526, #14673, #14240

**Pull request merged**